PresentData/PresentMonTraceConsumer.cpp

```diff
@@ -1139,6 +1188,169 @@ void PMTraceConsumer::HandleDWMEvent(EVENT_RECORD* pEventRecord)
+                // TODO: Assert that an element has been removed if this element must exist.
+            }
+        }
+    }
+
+    // mLastWindowPresent
+    {
+        // 0 is a invalid hwnd
+        if (p->Hwnd != 0) {
+            auto eventIter = mLastWindowPresent.find(p->Hwnd);
+            if (eventIter != mLastWindowPresent.end() && eventIter->second == p) {
+                mLastWindowPresent.erase(eventIter);
+            }
+            // There's no hard guaruntee that this element must exist, no assert needed here.
+        }
+    }
+
+    // mPresentsWaitingForDWM
+    {
+        if (p->WaitingForDwmCompletion)
+        {
+            auto eventIter = mPresentsWaitingForDWM.begin();
+            while (eventIter != mPresentsWaitingForDWM.end() && p != *eventIter)
+            {
+                // This loop should in theory be short because the present is old.
+                // If we are in this loop for dozens of times, something is likely wrong.
+                eventIter++;
+            }
+
+            if (eventIter != mPresentsWaitingForDWM.end())
+            {
+                mPresentsWaitingForDWM.erase(eventIter);
+            }
+            // Cannot assert that an event MUST be removed since it could be moved to some other present event's dependentPresents.
+        }
+    }
+
+    // mPresentsByLegacyBlitToken
+    {
+        // LegacyTokenData cannot be 0 if it's in mPresentsByLegacyBlitToken list.
+        if (p->LegacyBlitTokenData != 0) {
+            auto eventIter = mPresentsByLegacyBlitToken.find(p->LegacyBlitTokenData);
+            if (eventIter != mPresentsByLegacyBlitToken.end() && eventIter->second == p) {
+                mPresentsByLegacyBlitToken.erase(eventIter);
+                // TODO: Assert that an element has been removed if this element must exist.
+            }
+        }
+    }
+}
+
+void PMTraceConsumer::RemoveLostPresent(std::shared_ptr<PresentEvent> p)
+{
+    // This present has been timed out. Remove all references to it from all tracking structures.
+    // mPresentsByProcessAndSwapChain and mPresentsByProcess should always track the present's lifetime,
+    // so these also have an assert to validate this assumption.
+
+    DebugLostPresent(p);
+
+    // Only incomplete presents should make it here.
+    assert(!p->Completed);
+
+    // Remove the present from any struct that would only host the event temporarily.
+    // Should we loop through and remove the dependent presents?
+    RemovePresentFromTemporaryTrackingCollections(p);
+
+    // mPresentsByProcess
+    {
+        auto& presentsByThisProcess = mPresentsByProcess[p->ProcessId];
+        auto eventIter = presentsByThisProcess.find(p->QpcTime);
+        if (eventIter != presentsByThisProcess.end() && (eventIter->second == p)) {
+            presentsByThisProcess.erase(eventIter);
+            // TODO: Assert elementsRemoved = 1 if this element must exist. (What if QpcTime is not unique enough?)

```

Comment:
This seems like a good idea to me. Maybe just call the overload of `erase` that takes the key directly instead of doing a find first, and assert that the return value is <= 1.